Background reading: https://www.oreilly.com/learning/handling-missing-data

inspired by https://www.dataquest.io/blog/programming-best-practices-for-data-science/

data source: https://www.lendingclub.com/info/download-data.action

In [1]:
import time
import pandas
print(pandas.__version__)

0.23.4


In [2]:
pandas.options.display.max_rows = 999

In [3]:
start_time = time.time()
loans_2007 = pandas.read_csv('LoanStats3a.csv', skiprows=1, low_memory=False)
print(time.time() - start_time,'seconds')
loans_2007.shape

10.349263191223145 seconds


(42538, 145)

In [4]:
loans_2007.head(6).T

0  \
id                                                                                        NaN   
member_id                                                                                 NaN   
loan_amnt                                                                                5000   
funded_amnt                                                                              5000   
funded_amnt_inv                                                                          4975   
term                                                                                36 months   
int_rate                                                                               10.65%   
installment                                                                            162.87   
grade                                                                                       B   
sub_grade                                                                                  B2   
emp_title                                                                                 NaN   
emp_length                                                                          10+ years   
home_ownership                                                                           RENT   
annual_inc                                                                              24000   
verification_status                                                                  Verified   
issue_d                                                                              Dec-2011   
loan_status                                                                        Fully Paid   
pymnt_plan                                                                                  n   
url                                                                                       NaN   
desc                                          Borrower added on 12/22/11 > I need to upgra...   
purpose                                                                           credit_card   
title                                                                                Computer   
zip_code                                                                                860xx   
addr_state                                                                                 AZ   
dti                                                                                     27.65   
delinq_2yrs                                                                                 0   
earliest_cr_line                                                                     Jan-1985   
inq_last_6mths                                                                              1   
mths_since_last_delinq                                                                    NaN   
mths_since_last_record                                                                    NaN   
open_acc                                                                                    3   
pub_rec                                                                                     0   
revol_bal                                                                               13648   
revol_util                                                                              83.7%   
total_acc                                                                                   9   
initial_list_status                                                                         f   
out_prncp                                                                                   0   
out_prncp_inv                                                                               0   
total_pymnt                                                                           5863.16   
total_pymnt_inv                                                                       5833.84   
total_rec_prncp                                                                          5000   
total_rec_int     

## Empty columns

are there columns that have no values?

In [5]:
loans_2007.isnull().head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv   term  int_rate  \
0  True       True      False        False            False  False     False   
1  True       True      False        False            False  False     False   
2  True       True      False        False            False  False     False   
3  True       True      False        False            False  False     False   
4  True       True      False        False            False  False     False   

   installment  grade  sub_grade       ...         \
0        False  False      False       ...          
1        False  False      False       ...          
2        False  False      False       ...          
3        False  False      False       ...          
4        False  False      False       ...          

   hardship_payoff_balance_amount  hardship_last_payment_amount  \
0                            True                          True   
1                            True                          True   
2                            True                          True   
3                            True                          True   
4                            True                          True   

   disbursement_method  debt_settlement_flag  debt_settlement_flag_date  \
0                False                 False                       True   
1                False                 False                       True   
2                False                 False                       True   
3                False                 False                       True   
4                False                 False                       True   

   settlement_status  settlement_date  settlement_amount  \
0               True             True               True   
1               True             True               True   
2               True             True               True   
3               True             True               True   
4               True             True               True   

   settlement_percentage  settlement_term  
0                   True             True  
1                   True             True  
2                   True             True  
3                   True             True  
4                   True             True  

[5 rows x 145 columns]

In [6]:
loans_2007.isnull().sum()

id                                            42535
member_id                                     42538
loan_amnt                                         3
funded_amnt                                       3
funded_amnt_inv                                   3
term                                              3
int_rate                                          3
installment                                       3
grade                                             3
sub_grade                                         3
emp_title                                      2629
emp_length                                     1115
home_ownership                                    3
annual_inc                                        7
verification_status                               3
issue_d                                           3
loan_status                                       3
pymnt_plan                                        3
url                                           42538
desc        

In [7]:
loans_2007.shape[0]

42538

In [8]:
loans_2007.isnull().sum()==loans_2007.shape[0]

id                                            False
member_id                                      True
loan_amnt                                     False
funded_amnt                                   False
funded_amnt_inv                               False
term                                          False
int_rate                                      False
installment                                   False
grade                                         False
sub_grade                                     False
emp_title                                     False
emp_length                                    False
home_ownership                                False
annual_inc                                    False
verification_status                           False
issue_d                                       False
loan_status                                   False
pymnt_plan                                    False
url                                            True
desc        

Drop columns that are not useful

In [9]:
series_of_bool_indicating_whether_column_is_empty = loans_2007.isnull().sum()==loans_2007.shape[0]
loans_2007.shape

for col_name, bool_value in series_of_bool_indicating_whether_column_is_empty.iteritems():
    if bool_value:
        loans_2007.drop([col_name],axis=1,inplace=True)

In [10]:
loans_2007.shape

(42538, 64)

The original size was (42538, 145), so we deleted 81 columns.


For each column, what is the ratio of rows with missing entries to total row count?

In [11]:
len(loans_2007) # returns number of rows in dataframe

42538

In [12]:
loans_2007.isnull().sum()/len(loans_2007)

id                            0.999929
loan_amnt                     0.000071
funded_amnt                   0.000071
funded_amnt_inv               0.000071
term                          0.000071
int_rate                      0.000071
installment                   0.000071
grade                         0.000071
sub_grade                     0.000071
emp_title                     0.061804
emp_length                    0.026212
home_ownership                0.000071
annual_inc                    0.000165
verification_status           0.000071
issue_d                       0.000071
loan_status                   0.000071
pymnt_plan                    0.000071
desc                          0.312568
purpose                       0.000071
title                         0.000376
zip_code                      0.000071
addr_state                    0.000071
dti                           0.000071
delinq_2yrs                   0.000752
earliest_cr_line              0.000752
inq_last_6mths           

It would be better to view the series as a sorted sequence. 

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.sort_values.html

I wasn't aware of the existance of ".sorted_values()", but I figured it was a common task and <a href="https://www.google.com/search?q=pandas+sort+series">Google'd for the concept</a>.

In [13]:
(loans_2007.isnull().sum()/len(loans_2007)).sort_values(ascending=False)

id                            0.999929
settlement_status             0.996239
settlement_percentage         0.996239
debt_settlement_flag_date     0.996239
settlement_term               0.996239
settlement_date               0.996239
settlement_amount             0.996239
next_pymnt_d                  0.935375
mths_since_last_record        0.914171
mths_since_last_delinq        0.633058
desc                          0.312568
emp_title                     0.061804
pub_rec_bankruptcies          0.032159
emp_length                    0.026212
chargeoff_within_12_mths      0.003479
collections_12_mths_ex_med    0.003479
tax_liens                     0.002539
revol_util                    0.002186
last_pymnt_d                  0.002022
pub_rec                       0.000752
total_acc                     0.000752
delinq_amnt                   0.000752
delinq_2yrs                   0.000752
earliest_cr_line              0.000752
inq_last_6mths                0.000752
acc_now_delinq           

There are many columns that have 0.0071% of the entries containing NaN. How many rows is that?

In [14]:
loans_2007['total_rec_int'].isnull().sum()

3

The fact that many columns have the same number of NaNs is suspicious to me. To investigate, inspect those three rows

In [15]:
loans_2007[loans_2007['total_rec_int'].isnull()]

id  loan_amnt  \
39786         Loans that do not meet the credit policy        NaN   
42536  Total amount funded in policy code 1: 460296150        NaN   
42537          Total amount funded in policy code 2: 0        NaN   

       funded_amnt  funded_amnt_inv term int_rate  installment grade  \
39786          NaN              NaN  NaN      NaN          NaN   NaN   
42536          NaN              NaN  NaN      NaN          NaN   NaN   
42537          NaN              NaN  NaN      NaN          NaN   NaN   

      sub_grade emp_title       ...       tax_liens hardship_flag  \
39786       NaN       NaN       ...             NaN           NaN   
42536       NaN       NaN       ...             NaN           NaN   
42537       NaN       NaN       ...             NaN           NaN   

       disbursement_method debt_settlement_flag debt_settlement_flag_date  \
39786                  NaN                  NaN                       NaN   
42536                  NaN                  NaN                       NaN   
42537                  NaN                  NaN                       NaN   

      settlement_status settlement_date settlement_amount  \
39786               NaN             NaN               NaN   
42536               NaN             NaN               NaN   
42537               NaN             NaN               NaN   

      settlement_percentage settlement_term  
39786                   NaN             NaN  
42536                   NaN             NaN  
42537                   NaN             NaN  

[3 rows x 64 columns]

To summarize what I'm seeing, for three rows there is text in one column.

Let's count how many NaNs are in the 'id' column:

In [16]:
loans_2007['id'].isnull().sum()

42535

In [17]:
loans_2007.shape

(42538, 64)

The 'id' column is empty except for those three rows

My diagnosis is that the "id" column should not be part of the data analysis. 

The appropriate action would be to follow up with the person or team that supplied the data 

Before dropping the 'id' column and those three rows, let's validate that all other cells are NaN:

In [18]:
loans_2007[loans_2007['total_rec_int'].isnull()].isnull().sum().sum()

189

In [19]:
loans_2007[loans_2007['total_rec_int'].isnull()].shape

(3, 64)

In [20]:
3*64

192

Out of the 192 cells associated with the three rows, 189 cells are empty. That means we can safely drop both the "id" column and the three rows

In [21]:
print('before:',loans_2007.shape)

loans_2007.drop(['id'],axis=1,inplace=True)
loans_2007.dropna(how='all',axis=0,inplace=True) # drop empty rows

print('after:',loans_2007.shape)

before: (42538, 64)
after: (42535, 63)


# drop all rows and columns where entire row or column is NaN
If we wanted to be quick, we could have dropped all rows and columns containing NaNs

To show this, reload the data

In [22]:
start_time = time.time()
loans_2007 = pandas.read_csv('LoanStats3a.csv', skiprows=1, low_memory=False)
print(time.time() - start_time,'seconds')
loans_2007.shape

9.374513149261475 seconds


(42538, 145)

In [23]:
print('before:',loans_2007.shape)

loans_2007.dropna(how='all',axis=0,inplace=True) # rows 
loans_2007.dropna(how='all',axis=1,inplace=True) # columns

print('after:',loans_2007.shape)

before: (42538, 145)
after: (42538, 64)


The value of not taking the quick approach is that we gain understanding of the input data

## Are there columns where all rows have the same value?

In [24]:
for this_column in loans_2007.columns:
    if (loans_2007[this_column].nunique()==1):
        print(this_column)

pymnt_plan
initial_list_status
out_prncp
out_prncp_inv
collections_12_mths_ex_med
policy_code
application_type
chargeoff_within_12_mths
hardship_flag
disbursement_method


These columns probably aren't going to yield much useful information

In [25]:
print('before:',loans_2007.shape)

for this_column in loans_2007.columns:
    if (loans_2007[this_column].nunique()==1):
        loans_2007.drop([this_column],axis=1,inplace=True)
        
print('after:',loans_2007.shape)

before: (42538, 64)
after: (42538, 54)


We can inspect the remaining columns 

In [26]:
def uniq_per_col(df,count):
    for this_column in df.columns:
        print("==== ",this_column,"has",df[this_column].nunique(),"unique entries ====")
        print(df[this_column].value_counts().head(count))

In [27]:
uniq_per_col(loans_2007,5)

====  id has 3 unique entries ====
Total amount funded in policy code 2: 0            1
Loans that do not meet the credit policy           1
Total amount funded in policy code 1: 460296150    1
Name: id, dtype: int64
====  loan_amnt has 898 unique entries ====
10000.0    3016
12000.0    2439
5000.0     2260
6000.0     2037
15000.0    2012
Name: loan_amnt, dtype: int64
====  funded_amnt has 1051 unique entries ====
10000.0    2924
12000.0    2347
5000.0     2247
6000.0     2023
15000.0    1897
Name: funded_amnt, dtype: int64
====  funded_amnt_inv has 9248 unique entries ====
5000.0     1369
10000.0    1302
6000.0     1240
12000.0    1084
8000.0      929
Name: funded_amnt_inv, dtype: int64
====  term has 2 unique entries ====
 36 months    31534
 60 months    11001
Name: term, dtype: int64
====  int_rate has 394 unique entries ====
 10.99%    970
 11.49%    837
 13.49%    832
  7.51%    787
  7.88%    742
Name: int_rate, dtype: int64
====  installment has 16459 unique entries ====
311.11